### ISSUE TITLE
Sample performance

### ISSUE THREAD

**Issue by Yomguithereal** [simple-statistics/simple-statistics#250]
Currently, the `sample` function runs in `O(k)` (k being sample size) time but in `O(n)` (n being the size of the population). This can be awful for algorithms sampling large lists frequently (k-means, NN-descent etc...).

I developed some more efficients methods (having their drawback, however) in the [pandemonium](https://github.com/Yomguithereal/pandemonium) library (reservoir sampling still lacking but will soon be added). Does some of the library's method interest `simple-statistics`?

**Comment by tmcw** [comment]
Absolutely! Simple-statistics doesn't have much flexibility in sampling: I started with the Fisher-Yates approach because I could be comfortable with it being suitably random. Would love to add other methods, especially ones that are similarly random-enough and ones that work with streaming data.
